hydroflow_lang/src/graph/flat_graph_builder.rs

```diff
@@ -145,19 +223,109 @@ impl FlatGraphBuilder {
+                        Error::new(
+                            import.filename.span(),
+                            format!("filename: {}, err: {e}", import.filename.value()),
+                        )
+                    })?;
+
+                // TODO: see also above, parse_str sets all the spans in the resulting parsed token stream to the parent macro invocation span.
+                // This means that any error inside the imported module will manifest as a giant red squiggly line under the parent hydroflow_syntax!{} call.
+                let statements = match syn::parse_str::<HfCode>(&file_contents) {
+                    Ok(code) => code,
+                    Err(err) => {
+                        self.diagnostics.push(err.clone().into());
+                        return syn::Result::Err(err);
+                    }
+                };
+
+                let flat_graph_builder = crate::graph::FlatGraphBuilder::from_hfmodule(statements);
+                let (flat_graph, _uses, diagnostics) = flat_graph_builder.build();
+                diagnostics
+                    .iter()
+                    .for_each(crate::diagnostic::Diagnostic::emit);
+
+                self.merge_in(flat_graph, import.span())
+            }
+        }
+    }
+
+    /// Merge one flatgraph into the current flatgraph
+    /// other must be a flatgraph and not be partitioned yet.
+    fn merge_in(&mut self, other: HydroflowGraph, parent_span: Span) -> syn::Result<Ends> {
+        assert_eq!(other.subgraphs().count(), 0);
+
+        let mut ends = Ends {
+            inn: None,
+            out: None,
+        };
+
+        let mut node_mapping = BTreeMap::new();
+
+        for (nid, node) in other.nodes() {
+            match node {
+                Node::Operator(_) => {
+                    let varname = other.node_varname(nid);
+                    let new_id = self.flat_graph.insert_node(node.clone(), varname);
+                    node_mapping.insert(nid, new_id);
+                }
+                Node::ModuleBoundary { input, .. } => {
+                    let new_id = self.flat_graph.insert_node(
+                        Node::ModuleBoundary {
+                            input: *input,
+                            import_expr: parent_span,
+                        },
+                        Some(Ident::new(
+                            &format!("module_{}", input.to_string()),
+                            parent_span,
+                        )),
+                    );
+                    node_mapping.insert(nid, new_id);
+
+                    if *input {
+                        ends.inn =
+                            Some((PortIndexValue::Elided(None), GraphDet::Determined(new_id)));
+                    } else {
+                        ends.out =
+                            Some((PortIndexValue::Elided(None), GraphDet::Determined(new_id)));
+                    }
                 }
+                _ => panic!(),
```

Comment:
ought to fill these in each case with `Diagnostic` or `panic!()` messages for posterity